src/routes/CustomAsset/sagas/customAsset.js

```diff
@@ -1,151 +1,224 @@
 
-function* closeAddCustomAsset(): Saga<void> {
-  yield delay(config.delayBeforeFormClean)
-  yield put(clean())
-}
+const {
+  getContractName,
+  getContractSymbol,
+  getContractDecimals,
+} = web3
 
-function* addCustomAsset(): Saga<void> {
-  try {
-    const digitalAssets: DigitalAssets = yield select(selectDigitalAssetsItems)
-    const customAssetData: CustomAssetData = yield select(selectCustomAsset)
+type RequestedAssetFields = {|
+  decimals: number,
+  name: string,
+  symbol: string,
+|}
 
-    validate.customAssetData(customAssetData, digitalAssets)
-    const updatedAssets: DigitalAssets = digitalAssets.concat(getCustomAssetData(customAssetData))
+function addressFieldSimpleCheck(contractAddress: Address) {
+  const address = contractAddress.trim()
 
-    yield put(addSuccess(updatedAssets))
-  } catch (err) {
-    yield put(addError(err))
+  // 1. check length
+  if (address.length < 42) {
+    throw new InvalidFieldError('address', i18n('general.error.address.tooShort'))
   }
-}
 
-function* setEditAddress(action: { payload: { address: Address } }): Saga<void> {
-  try {
-    const digitalAssets: DigitalAssets = yield select(selectDigitalAssetsItems)
+  // 2. check address field checksum
+  if (!Keystore.checkAddressValid(address)) {
+    throw new InvalidFieldError('address', i18n('general.error.address.invalid'))
+  }
+}
 
-    const foundAsset: DigitalAsset = digitalAssets
-      .filter(({ address, isCustom }: DigitalAsset): boolean => {
-        if (address === action.payload.address) {
-          if (!isCustom) {
-            throw new Error('Digital asset is not custom')
-          }
+// -> to be moved to DigitalAssets saga
+function* getDigitalAsset(contractAddress: Address): Saga<DigitalAsset> {
+  const digitalAssets: DigitalAssets = yield select(selectDigitalAssetsItems)
 
-          return true
-        }
+  return digitalAssets.find(address => address === contractAddress)
+}
 
-        return false
-      })[0]
+/**
+ * Make requests to ETH node.
+ * Returns asset details or null if asset not ERC-20 compatible
+ *
+ * @param {string} contractAddress
+ * @returns {null | RequestedAssetFields}
+ */
+function* requestAssetFieldsTask(contractAddress: Address): Saga<?RequestedAssetFields> {
+  try {
+    // make three parallel requests
+    const { name, symbol, decimals } = yield all({
+      name: call(getContractName, contractAddress),
+      symbol: call(getContractSymbol, contractAddress),
+      decimals: call(getContractDecimals, contractAddress),
+    })
 
-    if (!foundAsset) {
-      throw new Error('Digital asset is not found')
+    if (name === '0x' || symbol === '0x') {
+      throw new Error('Invalid token')
     }
 
-    yield put(setEditAddressSuccess(foundAsset))
+    return {
+      name,
+      symbol,
+      decimals,
+    }
   } catch (err) {
-    yield put(setEditAddressError(err))
+    // catch Web3 network error
+    if (err.isOperational && err.cause.name !== 'BigNumber Error') {
```

Comment:
But we have to stop anyway, if error is catched

Reply:
Added check for Promise.TimeoutError here